File: POTD/Trie/Python/3093. Longest Common Suffix Queries.py
# // M- No of words in query
# // N = number of words in wordsContainer
# // Lcontainer=max len of word in container
# // Lquery=max len of word in query

# // TIME : O(N*Lcontainer + M*Lquery))
# // SPACE : O(N*Lcontainer + M)


class TrieNode:
    def __init__(self,ch):
        self.data=ch
        self.children=[None]*26
        self.isTerminalNode = False
        self.minWordIdx=-1

class Trie:
    def __init__(self):
        self.root=TrieNode('\0')
    
    def insertUtil(self,root, curIdx, word, wordsContainer, wordIdx):
        if root.minWordIdx==-1:
            root.minWordIdx=wordIdx
        else:
            curBestLen = len(wordsContainer[root.minWordIdx])
            newWordLen = len(word)

            if curBestLen > newWordLen:
                root.minWordIdx = wordIdx
        
        if curIdx==len(word):
            root.isTerminalNode=True
            return
        
        charIdx = ord(word[curIdx]) - ord('a')
        child=None

        if root.children[charIdx]:
            child=root.children[charIdx]
        else:
            child=TrieNode(word[curIdx])
            root.children[charIdx]=child
        
        self.insertUtil(child, curIdx + 1, word, wordsContainer, wordIdx)
    
    def insertWord(self, wordIdx, wordsContainer):
        self.insertUtil(self.root, 0, wordsContainer[wordIdx], wordsContainer, wordIdx)
    

    def searchWordIdx(self,query):
        cur=self.root

        for ch in query:
            charIdx=ord(ch)-ord('a')

            if cur.children[charIdx] is None:
                break
            else:
                cur=cur.children[charIdx]
        
        return cur.minWordIdx

class Solution:
    def stringIndices(self, wordsContainer: List[str], wordsQuery: List[str]) -> List[int]:

        for i in range(len(wordsContainer)):
            wordsContainer[i] = wordsContainer[i][::-1]

        for i in range(len(wordsQuery)):
            wordsQuery[i] = wordsQuery[i][::-1]
        
        t=Trie()

        for i in range(len(wordsContainer)):
            t.insertWord(i,wordsContainer)
        

        res = []

        for query in wordsQuery:
            res.append(t.searchWordIdx(query))
        
        return res
        

File: POTD/Trie/Cpp/3093. Longest Common Suffix Queries.cpp
// M- No of words in query
// N = number of words in wordsContainer
// Lcontainer=max len of word in container
// Lquery=max len of word in query

// TIME : O(N*Lcontainer + M*Lquery))
// SPACE : O(N*Lcontainer + M)


class TrieNode{
    public:
    char data;
    TrieNode* children[26];
    bool isTerminalNode;
    int minWordIdx;

    TrieNode(char ch){
        data=ch;
        for(int i=0;i<26;i++){
            children[i]=nullptr;
        }
        isTerminalNode=false;
        minWordIdx=-1;
    }


    ~TrieNode() {
        for (int i = 0; i < 26; i++) {
            if (children[i] != nullptr) {
                delete children[i];
            }
        }
    }
};


class Trie{
    public:
    TrieNode *root;

    Trie(){
        root=new TrieNode('\0');
    }

    ~Trie() {
        delete root;
    }
    void insertUtil(TrieNode *root,int curIdx,const string &word,const vector<string> &wordsContainer,int wordIdx){

        if(root->minWordIdx==-1){
            root->minWordIdx=wordIdx;
        }
        else{
            // the letter in the node should come from a string which is lesser in length or if equal should be present earlier in wordcontainer

            int curBestLen=wordsContainer[root->minWordIdx].length();
            int newWordLen=word.length();

            if(curBestLen>newWordLen){
                root->minWordIdx=wordIdx;
            }
        }


        if(curIdx==word.length()){
            root->isTerminalNode=true;
            return;
        }

        int charIdx=word[curIdx]-'a';
        TrieNode*child;

        if(root->children[charIdx]){
            child=root->children[charIdx];
        }
        else{
            child=new TrieNode(word[curIdx]);
            root->children[charIdx]=child;
        }

        insertUtil(child,curIdx+1,word,wordsContainer,wordIdx);
    }


    void insertWord(int wordIdx, const vector<string> &wordsContainer){
        insertUtil(root,0, wordsContainer[wordIdx],wordsContainer,wordIdx);
    }

    int searchWordIdx(const string &query){
        TrieNode*cur=root;

        for(char ch:query){

            if(cur->children[ch-'a']==nullptr){
                break;
            }
            else{
                cur=cur->children[ch-'a'];
            }
        }

        return cur->minWordIdx;
    }
};

class Solution {
public:
    vector<int> stringIndices(vector<string>& wordsContainer, vector<string>& wordsQuery) {

        for(auto &s:wordsContainer){
            reverse(s.begin(),s.end());
        }

        for (auto &q : wordsQuery) {
            reverse(q.begin(), q.end());
        }

        Trie t;

        for(int i=0;i<wordsContainer.size();i++){
            t.insertWord(i,wordsContainer);
        }
        
        vector<int> res;
        for(string &q:wordsQuery){
            res.push_back(t.searchWordIdx(q));
        }

        return res;
    }
};
